### ISSUE TITLE
Invalid cast when using inline "if" returning an enum

### ISSUE THREAD

**Issue by shaulbehr** [aspnet/EntityFrameworkCore#10293]
This one is easy to repro in EF Core 2.0:
```
var q = db.SomeTable
    .Select(x => new { EnumValue = x.SomeBoolean ? MyEnum.Yes : MyEnum.No })
    .ToList();
```
throws exception:
```
System.InvalidCastException : Specified cast is not valid.
   at lambda_method(Closure , Object[] )
   at Microsoft.EntityFrameworkCore.Storage.Internal.UntypedRelationalValueBufferFactory.Create(DbDataReader dataReader)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(Boolean buffer)
   at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```
This used to work in EF Core 1.1.0

**Comment by smitpatel** [comment]
This worked for me. Can you modify code below so that it demonstrate the exception you are seeing?
Tested using version 2.0.0 & 2.0.1
```C#
using System;
using System.Linq;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace EFSampleApp
{
    public class Program
    {
        public static void Main(string[] args)
        {
            using (var db = new MyContext())
            {
                // Recreate database
                db.Database.EnsureDeleted();
                db.Database.EnsureCreated();

                // Seed database
                db.AddRange(
                    new Blog { Value = true },
                    new Blog { Value = false }
                    );

                db.SaveChanges();
            }

            using (var db = new MyContext())
            {
                // Run queries
                var query = db.Blogs
                    .Select(e => new { Enum = e.Value ? MyEnum.Yes : MyEnum.No })
                    .ToList();
            }

            Console.WriteLine("Program finished.");
        }
    }


    public class MyContext : DbContext
    {
        private static ILoggerFactory LoggerFactory => new LoggerFactory().AddConsole(LogLevel.Trace);

        // Declare DBSets
        public DbSet<Blog> Blogs { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            // Select 1 provider
            optionsBuilder
                .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=_ModelApp;Trusted_Connection=True;Connect Timeout=5;ConnectRetryCount=0")
                //.UseSqlite("filename=_modelApp.db")
                //.UseInMemoryDatabase(databaseName: "_modelApp")
                .EnableSensitiveDataLogging()
                .UseLoggerFactory(LoggerFactory);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            // Configure model
        }
    }

    public class Blog
    {
        public int Id { get; set; }
        public bool Value { get; set; }
    }

    public enum MyEnum
    {
        Yes,
        No
    }
}

**Comment by shaulbehr** [comment]
@smitpatel Define `MyEnum` as a `byte`.
```
    public enum MyEnum : byte
    {
        Yes,
        No
    }
```

**Comment by smitpatel** [comment]
```
info: Microsoft.EntityFrameworkCore.Database.Command[200101]
      Executed DbCommand (1ms) [Parameters=[], CommandType='Text', CommandTimeout='30']
      SELECT CASE
          WHEN [e].[Value] = 1
          THEN 0 ELSE 1
      END AS [Enum]
      FROM [Blogs] AS [e]
fail: Microsoft.EntityFrameworkCore.Query[100100]
      An exception occurred in the database while iterating the results of a query for context type 'EFSampleApp.MyContext'.
      System.InvalidCastException: Unable to cast object of type 'System.Int32' to type 'System.Byte'.
         at lambda_method(Closure , Object[] )
         at Microsoft.EntityFrameworkCore.Storage.Internal.UntypedRelationalValueBufferFactory.Create(DbDataReader dataReader)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.BufferlessMoveNext(Boolean buffer)
         at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
         at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable`1.Enumerator.MoveNext()
         at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
```

**Comment by maumar** [comment]
fixed in 0027d598234866a8f3f844a64a8fe5f708173db6
